### ISSUE TITLE
Save `pyvista.MultiBlock` in STL format

### ISSUE THREAD

**Issue by scarpma** [pyvista/pyvista#2348]
Hello everyone.

I think that the option to save the `pyvista.MultiBlock` type in STL format would be helpful. Various software (e.g., Beta Ansa preprocessing tool for Computational Fluid Dynamics or Meshmixer) only accepts STL format for the input geometry. Multiple parts are allowed in the STL format by adding the `solid [solid_name] .... endsolid [solid_name]` keywords. This is an example (for STL ASCII, of course):
```
solid "front"
  facet normal -0.161384269595 0.98523491621 -0.0571600198746
    outer loop
      vertex 2.43927001953 -3.28291988373 31.002199173
      vertex 2.58858990669 -4.39023017883 31.0909004211
      vertex 3.39857983589 -3.73215985298 31.1854000092
    endloop
  endfacet
  facet normal -0.0869731009007 0.839549422264 -0.536276459694
    outer loop
      vertex 3.77996993065 1.47368001938 29.9706001282
      vertex 4.10769987106 2.22097992897 29.5272006989
      vertex 3.07201004028 2.18530011177 29.4426994324
    endloop
  endfacet
.........
endsolid "front"
solid "back"
  facet normal 0.348130375147 0.823633670807 0.447697281837
    outer loop
      vertex 10.0833997726 -10.0583000183 29.6856994629
      vertex 10.8774995804 -9.29601955414 29.7644004822
      vertex 9.88414001465 -9.12524032593 30.2770996094
    endloop
  endfacet
  facet normal -0.140576228499 0.74150198698 0.6560587883
    outer loop
      vertex 3.84086990356 -11.9118995667 28.7719993591
      vertex 3.13609004021 -11.3332004547 29.1504001617
      vertex 2.98072004318 -12.218000412 28.3381004333
    endloop
  endfacet
.........
endsolid "back"
```
I came up with a temporary solution using the already present function to save individual `pyvista.PolyData` in STL format:
```python
def save_multiblock_stl(multiblock, filename):

    names = multiblock.keys()
    oname, ext = osp.splitext(filename)
    assert ext == '.stl'
    
    # each individual stl file saved (output_filenames)
    ofiles = [f'{oname}_{ii}' + '.stl' for ii in range(len(names))]

    for ii, subpart in enumerate(multiblock):
        subpart.save(ofiles[ii], binary=False)
        change_first_line_of_file(ofiles[ii], f'solid {names[ii]}') # basically changes "solid" to "solid <solid_name>"

    # merge files together
    total_stl = ''
    for fn in ofiles:
        f = open(fn)
        total_stl += f.read()
        f.close()

    # writes total stl file
    with open(oname + '.stl', 'w') as f:
        f.write(total_stl)

    # deletes previously written stl files
    for fn in ofiles:
        os.remove(fn)

    return

def change_first_line_of_file(filename, new_first_line):

    fr = open(filename, 'r')
    first_line = fr.readline()
    fr.close()
    first_line_len = len(first_line)

    new_first_line_len = len(new_first_line)
    spaces_num = first_line_len - new_first_line_len
    new_first_line = new_first_line + ' '*(spaces_num-1) + '\n'
    fw = StringIO(new_first_line)
    fr = open(filename, 'r+')
    shutil.copyfileobj(fw, fr)
    fr.close()
    fw.close()
    return

```
Basically, I write to disk in STL ascii every single part of the `pyvista.MultiBlock`, I change the first line of each file from "solid" to "solid <solid_name>" and then I merge the files together. This works because the syntax
```
solid <solid_name>
...
...
endsolid
```
is allowed.

Obviously, this is not an efficient solution, so maybe someone has a better idea.

Thanks !

**Comment by banesullivan** [comment]
This is a fantastic idea! Thank you for the thorough explanation and sample code. I'm actually thinking this would be a great feature request upstream, directly in VTK. 

Would you be open to filing a feature request in https://gitlab.kitware.com/vtk/vtk ? Definitely ping me in the issue if you do

**Comment by scarpma** [comment]
Thank you for your reply! Actually, I'm not familiar with vtk alone. I could paste this same post in their forum at [https://discourse.vtk.org/c/development/7/l/latest](https://discourse.vtk.org/c/development/7/l/latest). What do you think?

**Comment by banesullivan** [comment]
Posting on the discourse forum would work too! Discourse might get more engagement

**Comment by hakostra** [comment]
I want to comment, that the "extension" of the STL by allowing multiple parts in one file with having several solid ... endsolid blocks are completely proprietary and non-standard. I'm not aware of anything else than ANSA that allows this as an extension.

I am myself a developer of a CFD software, and one of my worst nightmares are all various "extensions" to the ASCII STL file format. ANSA concatenates several ASCII files into one (as described), another CFD software preprocessor adds some "color" tags in addition to the "vertex" lines. Some STL writers happily writes "solid ...." in the 80 byte header of a binary STL, fooling quite a few readers into believing that the file is ASCII etc (this is strictly not forbidden, but imho. a very bad practice because it makes it hard to actually distinguish a binary STL from an ASCII STL).

Combined it becomes a nightmare to parse correctly. The closest thing to a written down STL standard is a [Wikipedia page](https://en.wikipedia.org/wiki/STL_(file_format)), which I think we should try to follow and not extend. It will just break half of the STL reader implementations out there...

So, please: don't do it.

**Comment by scarpma** [comment]
Thanks for the comment @hakostra . Actually also [Meshmixer](https://www.meshmixer.com/) reads this kind of files, but you're right.
In fact, neither pyvista splits the stl file according to the solid - endsolid convention.

@hakostra do you have suggestions for an open source and standard format for surface files for CFD preprocessors (like ANSA) ? Because the vtk format is not supported, for example. What formats do you use in your usual CFD pipeline ?

Thanks in advance

**Comment by hakostra** [comment]
I have never used Meshmixer. I have also no better suggestions. I'm just very annoyed by everyone inventing a new variant and extension. If you look in that  If we just could drop that ASCII STL file format and use binary STL in place it would be so much easier. An advantage is that they parse something like 100 times faster compared to ASCII. Then we just accept having a folder with a few hundred STL's describing individual parts.

This would actually be more in line with how VTK store it's collections (multiblock, AMR etc). There is one "parent" XML that just lists all "child" elements in the collection, that lie around in a subfolder. Imagine something like this `collection.vtm`:

```xml
<VTKFile type="vtkMultiBlockDataSet" version="1.0" byte_order="LittleEndian" header_type="UInt64">
  <vtkMultiBlockDataSet>
    <DataSet index="0" file="collection/front.stl"/>
    <DataSet index="1" file="collection/back.stl"/>
    <DataSet index="2" file="collection/window.stl"/>
    <DataSet index="3" file="collection/wheel.stl"/>
    <DataSet index="4" file="collection/door.stl"/>
  </vtkMultiBlockDataSet>
</VTKFile>
```
I'm quite sure it does not work right now, but it would be trivial to implement. It would not be supported by anything but VTK, though.

**Comment by scarpma** [comment]
I don't understand. What would not work ? Doesn't VTK already open .vtm files ? In what should this be implemented ?

**Comment by hakostra** [comment]
I don't know, haven't tried... I hand-wrote the example. Feel free to try it out if you like.

**Comment by scarpma** [comment]
I confirm that pyvista and vtk have functions to save .vtm files using this xml syntax you were mentioning.

My problem is that many CFD preprocessors do not have functions to open vtm files and so on.

If I understood correctly, your suggestion is to not modify vtk / pyvista libraries in order to fix problems of out of date softwares like ANSA.

**Comment by hakostra** [comment]
The problem is that everyone is doing their "variant" of the STL file format. Just try to open your "multiblock" STL in `numpy-stl`, and you will find it stops parsing after the first "endsolid", basically reading only the first part. And that's just the beginning.

Then Solidworks writes a 80 byte header in binary STL's thats not possible to distinguish from that of an ASCII: https://github.com/WoLpH/numpy-stl/issues/191

And just this single topic have been problematic for years:
https://github.com/WoLpH/numpy-stl/pull/80
https://github.com/WoLpH/numpy-stl/pull/157

And this is only a single reader implementation. I maintain my own reader in Fortran (because of lack of good alternatives), and I can promise it's a pain in the ass when someone invent something "clever" on their own. We once had an STL exported from StarCCM+ what contained a line "color" after the three "vertex" lines, just because StarCCM+ thought that was a great way to embed some color information. Guess what: it failed parsing.

In summary: the STL file format is a piece of junk. Mainly because there is no universally accepted standard. So then I follow these principles:

1. When writing an STL it shall be written in Binary STL format, according to the Wikipedia article.
2. When reading an STL, make the reader accept as much junk as possible. What you are interested in is just the collection of triangles. Ignore the rest. Also several solid..endsolid blocks. Merge all triangles and vertices to one part. This is more or less the VTK approach.

**Comment by scarpma** [comment]
Ok, you have been very clear. Thank you very much !

Just a question for curiosity. What does your reader in fortran do ? I mean, it parses ascii stl ? And then ? Does it convert to a specific data structure ?
